osu.Game/Screens/Play/HUD/KPSCounter/KeysPerSecondCalculator.cs

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable disable
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using osu.Framework.Timing;
+using osu.Game.Rulesets.Mods;
+using osu.Game.Rulesets.UI;
+
+namespace osu.Game.Screens.Play.HUD.KPSCounter
+{
+    public class KeysPerSecondCalculator : IDisposable
+    {
+        private static KeysPerSecondCalculator instance;
+
+        public static void AddInput()
+        {
+            instance?.onNewInput?.Invoke();
+        }
+
+        public static KeysPerSecondCalculator GetInstance(GameplayClock gameplayClock = null, DrawableRuleset drawableRuleset = null)
+        {
+            if (instance != null) return instance;
+
+            try
+            {
+                return new KeysPerSecondCalculator(gameplayClock, drawableRuleset);
+            }
+            catch (ArgumentNullException)
+            {
+                return null;
+            }
+        }
+
+        private readonly List<double> timestamps;
+        private readonly GameplayClock gameplayClock;
+        private readonly DrawableRuleset drawableRuleset;
+
+        private event Action onNewInput;
+
+        private IClock workingClock => (IClock)drawableRuleset?.FrameStableClock ?? gameplayClock;
```

Comment:
Fallback clock is removed in the next commit.
Also I need both clocks for accuracy:
- The `DrawableRuleset`'s `IFrameStableClock` is the clock that follows the sped up gameplay when seeking on a replay and since I need all inputs from the replay for backwards seeking, it allows me to add inputs while it's seeking forward. `GameplayClock`'s time follows what you hear and so directly skips to where you seeked, not ideal for what I want to achieve.
- The `GameplayClock` is used to get the gameplay rate and thus change the "1 second window" when a rate adjust mod is applied (and not get lower or higher values because the game is going too fast or too slow). `IFrameStableClock`'s rate takes into account user playback rate on replays, it would not make sense to have lower or higher values than expected because the user changed the replay playback rate.